File: src/game/items/BombItem.java
package game.items;

import game.Bomberman;
import game.entity.bomb.Bomb;
import game.events.game.UpdateCurrentAvailableBombsEvent;
import game.utils.Paths;
import game.utils.Utility;


public class BombItem extends UsableItem {
    private Bomb bombEntity;

    @Override
    public void use() {
        if (owner.getCurrExplosionLength() == 0) {
            return;
        }

        if (owner.getPlacedBombs() >= owner.getMaxBombs()) {
            return;
        }

        if (owner.getCurrentBombs() <= 0) {
            return;
        }

        if (Utility.timePassed(owner.getLastPlacedBombTime()) < Bomb.PLACE_INTERVAL) {
            return;
        }

        owner.setLastPlacedBombTime(System.currentTimeMillis());
        owner.setPlacedBombs(owner.getPlacedBombs() + 1);
        owner.setBombsSolid(false);

        new UpdateCurrentAvailableBombsEvent().invoke(owner.getCurrentBombs() - 1);
        bombEntity = new Bomb(owner);
        Bomberman.getMatch().addBomb(bombEntity);

        bombEntity.setOnExplodeListener(() -> {
            owner.setPlacedBombs(owner.getPlacedBombs() - 1);
            Bomberman.getMatch().removeBomb(bombEntity);
            new UpdateCurrentAvailableBombsEvent().invoke(owner.getCurrentBombs() + 1);
        });

        bombEntity.spawn(true);
        bombEntity.trigger();
    }

    @Override
    public String getImagePath() {
        return String.format("%s/bomb/bomb_0.png", Paths.getEntitiesFolder());
    }

    @Override
    public int getCount() {
        return owner.getCurrentBombs();
    }
}

File: src/game/items/BombItem.kt
package game.items

import game.Bomberman
import game.entity.placeable.Bomb
import game.events.game.UpdateCurrentAvailableItemsEvent
import game.utils.Paths.entitiesFolder
import game.utils.Utility.timePassed

class BombItem : UsableItem() {
    private lateinit var bombEntity: Bomb

    override fun use() {
        val match = Bomberman.getMatch() ?: return

        val isLocalPlayer = owner == match.player
        val isBombPlacementValid = isLocalPlayer &&
                (owner.currExplosionLength == 0 || owner.placedBombs >= owner.maxBombs || owner.currentBombs <= 0)

        val isBombPlacementIntervalValid = timePassed(owner.lastPlacedBombTime) < Bomb.PLACE_INTERVAL

        if (isBombPlacementValid || isBombPlacementIntervalValid) {
            return
        }

        owner.lastPlacedBombTime = System.currentTimeMillis()
        owner.placedBombs++
        owner.setBombsSolid(false)

        if (isLocalPlayer)
            UpdateCurrentAvailableItemsEvent().invoke(owner.currentBombs - 1)

        bombEntity = Bomb(owner)

        match.addBomb(bombEntity)

        bombEntity.onExplodeCallback = {
            owner.placedBombs--
            match.removeBomb(bombEntity)

            if (isLocalPlayer)
                UpdateCurrentAvailableItemsEvent().invoke(owner.currentBombs + 1)
        }

        bombEntity.spawn(true)
        bombEntity.trigger()
    }

    override fun combineItems(item: UsableItem) {
    }


    override val imagePath: String = "$entitiesFolder/bomb/bomb_0.png"

    override val count: Int
        get() = owner.currentBombs

    override val type: ItemsTypes = ItemsTypes.BombItem
}